### PR TITLE
vector-attributes-in-descriptor

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Vektor-Attribute im ResultDescriptor](https://trello.com/c/Kus16UbI/205-vektor-attribute-im-resultdescriptor)

File: operators/src/engine/result_descriptor.rs [hunk 1]
```diff
@@ -69,10 +71,25 @@ impl ResultDescriptor for RasterResultDescriptor {
...
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
 pub struct VectorResultDescriptor {
     pub data_type: VectorDataType,
     pub spatial_reference: SpatialReferenceOption,
+    pub columns: HashMap<String, FeatureDataType>,
+}
+
+impl VectorResultDescriptor {
+    /// Create a new `VectorResultDescriptor` by only using new `columns`
+    pub fn map_columns<F>(&self, f: F) -> Self
```
Comment (michaelmattig) [inline operators/src/engine/result_descriptor.rs:83]: why not consume self like the other map functions on result descriptor?

File: operators/src/engine/result_descriptor.rs [hunk 2]
```diff
@@ -69,10 +71,25 @@ impl ResultDescriptor for RasterResultDescriptor {
...
-#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
 pub struct VectorResultDescriptor {
     pub data_type: VectorDataType,
     pub spatial_reference: SpatialReferenceOption,
+    pub columns: HashMap<String, FeatureDataType>,
+}
+
+impl VectorResultDescriptor {
+    /// Create a new `VectorResultDescriptor` by only using new `columns`
```
Comment (michaelmattig) [inline operators/src/engine/result_descriptor.rs:82]: why is columns in tick when there is no parameter named columns? Also why do they all have to be new? Then I wouldn't use map but new

File: operators/src/mock/mock_raster_point_join.rs [hunk 3]
```diff
@@ -91,23 +92,40 @@ impl VectorOperator for MockRasterPointJoinOperator {
...
+
+        let raster_sources = self
+            .raster_sources
+            .into_iter()
+            .map(|o| o.initialize(context))
+            .collect::<Result<Vec<_>>>()?;
+
+        let result_descriptor = {
+            let mut columns = vector_sources[0].result_descriptor().columns.clone();
+            columns.insert(self.params.feature_name.clone(), FeatureDataType::Number);
```
Comment (michaelmattig) [inline operators/src/mock/mock_raster_point_join.rs:109]: check if column with that name already exists?

File: operators/src/mock/mock_raster_point_join.rs [hunk 4]
```diff
@@ -91,23 +92,40 @@ impl VectorOperator for MockRasterPointJoinOperator {
...
+
+            VectorResultDescriptor {
+                spatial_reference: vector_sources.get(0).map_or_else(
```
Comment (michaelmattig) [inline operators/src/mock/mock_raster_point_join.rs:112]: I think you can just unwrap here because you used the first vector source already so it has to exist

File: operators/src/mock/mock_raster_point_join.rs [hunk 5]
```diff
@@ -91,23 +92,40 @@ impl VectorOperator for MockRasterPointJoinOperator {
...
+            .collect::<Result<Vec<_>>>()?;
+
+        let raster_sources = self
+            .raster_sources
+            .into_iter()
+            .map(|o| o.initialize(context))
+            .collect::<Result<Vec<_>>>()?;
+
+        let result_descriptor = {
+            let mut columns = vector_sources[0].result_descriptor().columns.clone();
```
Comment (michaelmattig) [inline operators/src/mock/mock_raster_point_join.rs:108]: check if a raster source exists

File: operators/src/mock/mock_raster_point_join.rs [hunk 6]
```diff
@@ -91,23 +92,40 @@ impl VectorOperator for MockRasterPointJoinOperator {
...
+            columns.insert(self.params.feature_name.clone(), FeatureDataType::Number);
+
+            VectorResultDescriptor {
+                spatial_reference: vector_sources.get(0).map_or_else(
+                    || SpatialReferenceOption::Unreferenced,
+                    |o| o.result_descriptor().spatial_reference,
+                ),
+                data_type: VectorDataType::MultiPoint,
```
Comment (michaelmattig) [inline operators/src/mock/mock_raster_point_join.rs:116]: do we need to check that first vector source is actually MultiPoint?

File: operators/src/processing/raster_vector_join/mod.rs [hunk 7]
```diff
@@ -86,14 +88,40 @@ impl VectorOperator for RasterVectorJoin {
         // TODO: check for column clashes earlier with the result descriptor
         // TODO: update result descriptor with new column(s)
```
Comment (michaelmattig) [inline operators/src/processing/raster_vector_join/mod.rs:89]: remove

File: operators/src/processing/raster_vector_join/mod.rs [hunk 8]
```diff
@@ -86,14 +88,40 @@ impl VectorOperator for RasterVectorJoin {
...
+                    AggregationMethod::First => {
+                        match raster_sources[i].result_descriptor().data_type {
+                            RasterDataType::U8
+                            | RasterDataType::U16
+                            | RasterDataType::U32
+                            | RasterDataType::U64
+                            | RasterDataType::I8
+                            | RasterDataType::I16
+                            | RasterDataType::I32
+                            | RasterDataType::I64 => FeatureDataType::Decimal,
```
Comment (michaelmattig) [inline operators/src/processing/raster_vector_join/mod.rs:110]: remark: I still find it confusing that our integer types are named Decimal when e.g. in Postgres Decimal is a variable precision data type with a decimal point.

File: operators/src/processing/vector_join/util.rs [hunk 9]
```diff
@@ -0,0 +1,22 @@
+use std::collections::{HashMap, HashSet};
+
+/// Create a translation table to resolve name conflicts in the `DataCollection`
+pub(super) fn translation_table<'i>(
```
Comment (michaelmattig) [inline operators/src/processing/vector_join/util.rs:4]: would be a good idea to test this function standalone, especially with the edge cases of suffixeing stuff multiple times as long as collisions occur

File: services/src/handlers/workflows.rs [hunk 10]
```diff
@@ -49,6 +52,37 @@ async fn load_workflow<C: Context>(id: Uuid, ctx: C) -> Result<impl warp::Reply,
     Ok(warp::reply::json(&wf).into_response())
 }
 
+pub(crate) fn get_workflow_metadata_handler<C: Context>(
+    ctx: C,
+) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+    warp::get()
+        .and(warp::path!("workflow" / "metadata" / Uuid))
```
Comment (michaelmattig) [inline services/src/handlers/workflows.rs:59]: wouldn't it be more RESTful to call /workflow/UUid/metdata ?
and maybe /workflow/UUid/definition for the actual workflow? then /workflow/UUid could deliver both..

File: services/src/handlers/workflows.rs [hunk 11]
```diff
@@ -49,6 +52,37 @@ async fn load_workflow<C: Context>(id: Uuid, ctx: C) -> Result<impl warp::Reply,
...
+    ctx: C,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    // TODO: check for rights to access workflow?
+    let workflow = ctx
+        .workflow_registry_ref()
+        .await
+        .load(&WorkflowId(id))
+        .await?;
+    let operator = workflow.operator.get_vector().context(error::Operator)?;
+    // TODO: use meaningful context to initialize
```
Comment (michaelmattig) [inline services/src/handlers/workflows.rs:75]: you can tet it from `ctx` now

File: services/src/handlers/workflows.rs [hunk 12]
```diff
@@ -49,6 +52,37 @@ async fn load_workflow<C: Context>(id: Uuid, ctx: C) -> Result<impl warp::Reply,
...
+    let operator = operator
+        .initialize(&MockExecutionContext::default())
+        .context(error::Operator)?;
+
+    let result_descriptor = operator.result_descriptor();
```
Comment (michaelmattig) [inline services/src/handlers/workflows.rs:80]: we should think about caching this. maybe we can calculate it once on workflow creation? It doesn't depend on the query rectangle.

File: operators/src/engine/result_descriptor.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply operators/src/engine/result_descriptor.rs:83]: Because you now get result descriptors by reference and then you would always need to say clone and then map. Then you've cloned data and discard it, e.g. a hash map, so I think it makes sense this way.

File: operators/src/processing/raster_vector_join/mod.rs (hunk 8 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/raster_vector_join/mod.rs:110]: True, maybe we can sit together and think of reworks to the feature data types for a small story.

File: operators/src/engine/result_descriptor.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply operators/src/engine/result_descriptor.rs]: ba5480b 

You don't have to, you only get a reference of the hash map.

File: operators/src/mock/mock_raster_point_join.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply operators/src/mock/mock_raster_point_join.rs]: 1b9537d

File: operators/src/mock/mock_raster_point_join.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply operators/src/mock/mock_raster_point_join.rs:134]: 1b9537d

File: operators/src/mock/mock_raster_point_join.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply operators/src/mock/mock_raster_point_join.rs:150]: 1b9537d

File: operators/src/processing/raster_vector_join/mod.rs (hunk 7 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/raster_vector_join/mod.rs]: 1d3aea5

File: operators/src/processing/vector_join/util.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/vector_join/util.rs:4]: 5477481

File: services/src/handlers/workflows.rs (hunk 10 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/workflows.rs]: 7075bd4

File: services/src/handlers/workflows.rs (hunk 12 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/workflows.rs:81]: I added a todo

7075bd4 

caching it would mean we compiled it ;-)

File: services/src/handlers/workflows.rs (hunk 11 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/workflows.rs]: afbbacb

File: operators/src/mock/mock_raster_point_join.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply operators/src/mock/mock_raster_point_join.rs]: 69e3441
